README.md
# CriptoSapo

## A brand new never ending project... that now has an end???

Adaptado a la gran linea, el antiguo protagonista, sapo, se ha
convertido en Usopp y fue sustituido por Luffy.

Ahora este proyecto es un reto para aprender a usar herramientas que
permitan desarrollar programas faciles y utiles, los cuales, en el
futuro, me salvaran el pellejo.

### Caracteristicas de la version 1

- Registrar una base de datos en vivo desde binance.
- Construir grafico de velas usado en trading.
- Simular compra y venta en vivo de BTC con respecto al USDT.
- Interfaz grafica para
	- Capturar datos;
	- Transformar / exportar datos;
	- Simular compra;
	- Simular venta.
- Grafico para completar proceso ETL (queryem).
- Standalone.

### Proximamente...

Camino a la version 2 ~si es que no se abandona el proyecto~ se
buscara una version multiplataforma, modernizar la etapa de
compilacion y agregar un selector de tokens. Lo anterior sera
respaldado eventualmente por una lista de tareas WIP__...__


changelog.md
# The third and final / first version!!!

`build date: 2022-10-11 22:56:07`
`build version: 1.m4A1`

- added front page.
	- added 4 buttons.
	- styled the whole page.
	- added icon.
	- added javascript file for making wires.
- readme typos fixed.
- added application launcher, `app.bat`.
- Robin has joined the chat.
	- Robin is the main server.
	- added API url, 'wallet', for listening to all CRUD actions.
	- added GET action to display the wallet's content.
- modified all `.bat` files to exit at the end of file.
- updated look and feel of home page.
- added css classes to show unavailable buttons.
- made API response actions for 'buy' and 'sell' buttons.
- say hello to Sanji, the Robin's dog!
	- sets the current directory to 'action'.
	- returns the wallet's content.
- made API test calls.
	- added Insomnia API test json and yaml files.
- deleted yaml file.
- added 'delete wallet' button.
- updated links to static files (css, js, png...).
- added backend functions for 'ship' and 'spy' buttons.
- created API calls in frontend except for 'del' button.
- api unit tests updated.
- clicker added.
- updated readme with promises.
- added to the build instructions
	- server Robin;
	- app launcher;
	- server dependencies;
	- server resources.

---

# Standalone is here!!

`build date: 2022-09-30 17:10:57`
`build version: 0.m4p4`

- trying to remove glitch from changelog. Let's see next push...
- working with branches from now on.
- ignored build zip files.
- fixed changelog markdown breakline.
- `init.bat` updated to evaluate env vars which add the current python dist to the path.
	- added `p` parameter to launch python interpreter.
- activated git LFS for the `.xlsx` files.
- updated readme to indicate the ETL process in the `view.xlsx` file (done at the very beginning of the project).
	- updated queryem to be independent of file path.
- cleaned up the `.xlsx` file.
- fixed typo in previous commit for future changelogs.
- modified everything to embed python.
	- added python distribution folder to app folder.
	- updated `.gitignore` to ignore every python distribution.
	- updated bat files (buy, sell, ship, spy) to work with embedded python.
- updated json indentation inside chauchera.
- updated copy function in `build.py` to:
	 - ignore comments.
	 - ignore empty lines.
- added pydist to build folder.
- added python distribution path to `src.txt`. 
- readme updated to celebrate new release.

---

# Build tools out of the box, 2/2! Keep moving from now on and go in peace.

`build date: 2022-09-12 00:48:01`
`build version: 0.m473`

- added app info file
- added `src.txt` file to read source files to be copied
- make build tools
	- erase last build content
	- check if build folder set up
	- copy new project content
	- versioning system
		- number base convertor
	- added beta versions, which does't update changelog on build
	- update changelog
		- changelog renamed from 'txt' to 'md'
	- make zip distribution
	- reordered build instructions
- make changelog file
- updated format issue in readme
- pydist ignored
- new project folders structure (files moved to src folder)
- removed trash from `del_build_content()` and added comments
- removed commented import lines in `build.py`
- readme updated
- changed `b.bat` to `build.bat`
- I screw it up. It'll not be 11 but 2 weeks instead v:
- __first public release with a tag!!__ _When I find out how to upload it..._

---

# Build tools out of the box, 1/2! New doggo's toy.

date: 2022-08-30 00:07:52

- added app info file \
.- make build tools
    - erase last build content
    - check if build folder set up \
    .- copy new project content \
    .- versioning system
        - number base convertor \
    .- update changelog \
    .- make zip distribution \
.- make changelog file
- updated format issue in readme
- pydist and bman ignored

---

# Everyone stand up in the Going Merry!!

date: 2022-08-21 14:48:42

- The general mini on board refactor
    - everyone moved to going_merry
    - nami kissed sapo and it converted into... usopp?
    - sapo
        - renamed to usopp
            - say hello to usopp!!
        - old schema function, pAPP(), renamed to spy()
    - nami dependencies updated from sapo to usopp
    - luffy dependencies updated from sapo to usopp
    - dev tools updated to fit in going_merry
    - action buttons updated to fit in going_merry
- now nami says when no wallet was created
- franky
    - updated format time function to optionally give seconds
    - franky's franky renamed to build_blueprint()
- blazingly new readme to meet the actual purpose of the project
- updated dev tools to inject ingored files into app's folder
- added comment to every button indicating python's path
- nami now checks if the deposited amount is valid (int > 0)
- now nami saves the seconds of trades in wallet's history
- removed some dumb spaces

---

# Say hello to Luffitaro

date: 2022-08-20 19:46:55

- body of main function
    - logic flow and modules imported
    - orders to buy
    - orders to sell
    - orders to spy
    - orders to ship
    - bat file buttons for every order
- luffy has a whip

Other mini goals:
- nami now verifies if coin index exists in wallet when selling
- nami's chat is now capitalized
- sapo's chat is now capitalized and pprint removed
- franky's chat is now capitalized and pprint removed
- changed wallet database url in nami
- added del new_trade statemant in nami
- changed databases url in franky and sapo
- changed 'select' button
- added 'init.bat' to add dev tools to path in command line
- changed 't.bat' to fit new path
- added database verification function to franky
- added description to make_format_time() in franky
- added a dot to the end of every function description

---

# say hello to Onami

date: 2022-08-17 23:00:09

- wallet functions added
    - json structure
    - write to
    - read from
    - update pocket
    - first balance option
- wallet gitignored

tools/build.py
import os
from sys import exit as build_notification, argv
import json
import time
from zipfile import ZipFile as zfile, ZIP_DEFLATED


def del_build_content() -> None:
     ("Recorre toda la carpeta 'build' para eliminar su contenido.")
     build_folder_tree = list(os.walk('.\\build'))[::-1]
     for path, folders, files in build_folder_tree:
          for file in files:
               print(path + '\\' + file)
               os.unlink(path + '\\' + file)
          for folder in folders:
               print(path + '\\' + folder)
               os.rmdir(path + '\\' + folder)
     return None


def check_build_folder(project_name: str) -> bool:
     ("Verifica si existe la carpeta 'build'. Si no existe, intenta"
     " crearla. Solo retorna True si, al verificar, no encuentra"
     " errores ni necesita crear carpetas o archivos")
     if os.getcwd().endswith(project_name):
          try:
               ITER_AGAIN = False
               if not 'build' in os.listdir():
                    os.mkdir('build')
                    ITER_AGAIN = True
               if not 'pydist' in os.listdir('tools'):
                    os.mkdir('tools\\pydist')
                    ITER_AGAIN = True
               if not 'bdmn.txt' in os.listdir('tools'):
                    legos = {
                         "app name": '',
                         "app version": '0',
                         "beta version": '1',
                         "build date": '0',
                         "build number": '0',
                         "build version": '0',
                         "changelog file": '',
                         "commit file": '',
                         "license": '',
                         "python version": '',
                    }
                    with open('tools\\bdmn.txt', 'x') as info_file:
                         json.dump(legos, info_file, indent=2)
                    ITER_AGAIN = True
          except Exception as e:
               print('A')
               print(e)
               return False
          if ITER_AGAIN: return False
     else:
          print(f'No estas en la carpeta correcta del proyecto {project_name}')
          return False
     return True


def check_app_info() -> bool:
     if os.path.isfile('tools\\bdmn.txt'):
          with open('tools\\bdmn.txt') as info_file:
               legos: dict = json.load(info_file)
     else:
          return False

     print()
     for info in legos:
          print(info + ': ' + legos[info])
          if legos[info] == '':
               print('No puede haber campos vacios.')
               return False
     print()

     if legos['app version'].isnumeric():
          if int(legos['build number']) < 0:
               print('Major version must be a possitive integer')
               return False
     else:
          print('Build number must be a possitive integer')
          return False

     if legos['beta version'].isnumeric():
          if not int(legos['beta version']) in (0, 1):
               print('Beta version must be a 0 or 1')
               return False               
     else:
          print('Beta version must be 0 or 1')
          return False

     if legos['build number'].isnumeric():
          if int(legos['build number']) < 0:
               print('Build number must be a possitive integer')
               return False
     else:
          print('Build number must be a possitive integer')
          return False

     if not os.path.isfile(legos['changelog file']):
          print('No se puede encontrar changelog file.')
          return False
     if not os.path.isfile(legos['commit file']):
          print('No se puede encontrar commit file.')
          return False

     python_version = legos['python version'].split('.')

     if len(python_version) != 2:
          print('Version de python mal indicada.')
          return False
     elif not python_version[0].isnumeric():
          print('Version de python mal indicada.')
          return False
     elif not python_version[1].isnumeric():
          print('Version de python mal indicada.')
          return False

     return True


def klaus_fangs(x: str, base_a: int, base_b: int) -> str:
     ("Convierte un numero 'x' en base 'a' al equivalente en base"
     " 'b'. 'x' debe ser un entero positivo y las bases van desde"
     " base 2 hasta base 62 inclusive.")
     if isinstance(x, int): x = str(x)
     if not isinstance(x, str): raise TypeError
     if not isinstance(base_a, int): raise TypeError
     if not isinstance(base_b, int): raise TypeError
     if not 1 < base_a < 63: raise ValueError
     if not 1 < base_b < 63: raise ValueError
     decimal_of = {
          '0': 0,  '1': 1,  '2': 2,  '3': 3,  '4': 4,  '5': 5,  '6': 6,  '7': 7,  '8': 8,  '9': 9,  'a': 10,
          'b': 11, 'c': 12, 'd': 13, 'e': 14, 'f': 15, 'g': 16, 'h': 17, 'i': 18, 'j': 19, 'k': 20, 'l': 21,
          'm': 22, 'n': 23, 'o': 24, 'p': 25, 'q': 26, 'r': 27, 's': 28, 't': 29, 'u': 30, 'v': 31, 'w': 32,
          'x': 33, 'y': 34, 'z': 35, 'A': 36, 'B': 37, 'C': 38, 'D': 39, 'E': 40, 'F': 41, 'G': 42, 'H': 43,
          'I': 44, 'J': 45, 'K': 46, 'L': 47, 'M': 48, 'N': 49, 'O': 50, 'P': 51, 'Q': 52, 'R': 53, 'S': 54,
          'T': 55, 'U': 56, 'V': 57, 'W': 58, 'X': 59, 'Y': 60, 'Z': 61,
     }
     for letra in x:
          if not letra in decimal_of.keys():
               raise ValueError
     # from base a to decimal
     y = 0
     posicion = len(x) - 1
     for x0 in x:
          if base_a < decimal_of[x0] + 1: raise TypeError
          y += decimal_of[x0]*base_a**posicion
          posicion -= 1
     # from decimal to base b
     X = ''
     while ++y:
          if y < base_b:
               X += list(decimal_of.items())[y][0]
               break
          X += list(decimal_of.items())[y % base_b][0]
          y = y//base_b
     if X == '': X = '0'
     return X[::-1]


def version_it() -> str:
     ("Updates build number and its date inside app's information file.")
     legos: dict
     fecha: [str, str]
     build_number: str
     version_code: str
     with open('tools\\bdmn.txt') as info_file:
          legos = json.load(info_file)
          new_build_date: str = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime())
          fecha = time.strftime('%y.%j', time.localtime()).split('.')
          build_number = str(int(legos['build number']) + 1)
          version_code = klaus_fangs(fecha[0], 10, 62) \
                       + klaus_fangs(fecha[1], 10, 62) \
                       + klaus_fangs(build_number, 10, 62)
     with open('tools\\bdmn.txt', 'w') as info_file:
          legos['build date'] = new_build_date
          legos['build number'] = build_number
          legos['build version'] = legos['app version'] + '.' + version_code
          print(json.dumps(legos, indent=2))
          json.dump(legos, info_file, indent=2)
     return legos['build version']


def update_changelog() -> None:
     ("Writes on a changelog file defined in the app information file"
     " the content of the commit file, adding the date and build numbers."
     " It only do the writing if the build is not a beta version.")
     with open('tools\\bdmn.txt') as info_file:
          legos: dict = json.load(info_file)
          if int(legos['beta version']): return None
          # retrieving information
          with open(legos['commit file']) as commit_file:
               commit_content = commit_file.readlines()
               commit_header = '# ' + commit_content.pop(0)
               commit_content = ''.join(commit_content)
          with open(legos['changelog file']) as changelog_file:
               changelog_content = changelog_file.read()
          # merging and updating information into changelog
          with open(legos['changelog file'], 'w') as changelog_file:
               new_changelog_content = commit_header          \
                                     + '\n`build date: '      \
                                     + legos['build date']    \
                                     + '`\n`build version: '  \
                                     + legos['build version'] \
                                     + '`\n'                  \
                                     + commit_content         \
                                     + '\n---\n\n'            \
                                     + changelog_content
               changelog_file.write(new_changelog_content)
     return None


def copy_sources() -> bool:
     ("Copy the specified source files in 'src.txt' to the build folder.")

     if not os.path.isfile('tools/src.txt'):
          print('Sin archivo \'src.txt\'')
          return False

     with open('tools/src.txt') as sources:
          source_files = sources.readlines()

     with open('tools/_src.txt', 'w') as xcopy:
          for file in source_files:
               if file == '\n': continue
               if file[0] == '#': continue
               file = file[:-1:]
               destination = file.split('\\')
               destination.pop()
               if destination and destination[0] == 'app':
                    destination.pop(0)
               if destination and destination[0] == 'tools':
                    destination.pop(0)
               command = 'xcopy ' + file + ' build\\' + '\\'.join(destination) + '\n'
               xcopy.write(command)

               # crea las carpetas paa que xcopy no se maree
               last_folder = 'build'
               for folder in destination:
                    if not os.path.isdir(last_folder + '/' + folder):
                         os.mkdir(last_folder + '/' + folder)
                    last_folder += '/' + folder

     return True


def compile_project() -> None:
     ("Compile .py files to .pyc files.")
     pass


def pack_it() -> None:
     ("Packs build folder into a zip file.")

     with open('tools\\bdmn.txt') as info_file:
          legos: dict = json.load(info_file)

     os.chdir('build')
     source_files = list(os.walk('.'))
     app_name: str = f"criptosapo-{'-'.join(legos['build version'].split('.'))}.zip"

     with zfile(
          app_name,
          mode = 'w',
          compression = ZIP_DEFLATED,
          compresslevel = 9
     ) as app:
          for path, folders, files in source_files:
               for folder in folders:
                    print(path + '\\' + folder)
                    app.write(path + '\\' + folder)
               for file in files:
                    print(path + '\\' + file)
                    app.write(path + '\\' + file)

     os.chdir('..')

     return None


def build() -> None:
     PROJECT_NAME = 'CriptoSapo'
     INSTRUCTION: str = argv[-1::][0] if len(argv) > 1 else ''
     NOT_READY, UPDATE, COPY, BUILD = 1, 4, 33, 69

     match INSTRUCTION:
          case 'setup':
               print('Checking files and setting up...')
               del_build_content()
               FOLDER_READY = check_build_folder(PROJECT_NAME)
               if not FOLDER_READY:
                    return print(
                         'Build folder initialized.'
                         ' Update your config file.'
                    )
               if not check_app_info():
                    build_notification(NOT_READY)
               build_notification(UPDATE)
          case 'update':
               BUILD_VERSION: str = version_it()
               print('VASDFasdfbaksfbkasef:', BUILD_VERSION)
               update_changelog()
               build_notification(COPY)
          case 'copy':
               if copy_sources(): # done in batch for now
                    build_notification(BUILD)
          case 'build':
               print('Building...')
               compile_project()
               pack_it()
     return None


build()

#ned
